### DISCUSSION TITLE
Advice on ensuring safe queries

### DISCUSSION THREAD

**Discussion by liquidcarbon** [duckdb/duckdb#11634] [Q&A]
Hello,

I'm building a web UI with FastAPI that allows users to run queries against local files via DuckDB.
The server and web UI live behind corporate VPN.

Though I don't anticipate any SQL injections or other shenanigans from internal teams, I'd like to ask for best practices that would ensure safe queries.

The query constructor I like to use looks like this:

```python
class Q(str):
    UNSAFE = ["CREATE", "DELETE", "DROP", "INSERT", "UPDATE"]
    
    def __init__(self, query: str):
        self.is_safe = not any([cmd in query.upper() for cmd in self.UNSAFE])

    def run_duckdb(self):
        return duckdb.sql(self)
```

The query is constructed through a web form and executes when `.is_safe` is True.

I don't think I can or want to prevent a user requesting a gazillion rows if a join goes wrong or some recursive queries; would rather have a query time out after, say, 15 min.

Does this seem like a (minimally) reasonable approach?

**Comment by Tishj** [comment]
On the C++ side there's an ExtractStatements method that parses the query string into Statement(s)

We recently added it to the Python API under `extract_statements`
You can use that to figure out types of the statements
